src/protocols/secure_channel/CASESession.cpp

```diff
@@ -203,6 +203,27 @@ class CASESession::WorkHelper
 
     bool IsCancelled() const { return mSession.load() == nullptr; }
 
+    // This API returns true when background thread fails to schedule the AfterWorkCallback
+    bool SchedulingAfterWorkCallbackFailed() { return mScheduleAfterWorkFailed.load(); }
+
+    // This API runs the AfterWorkCallback. This is supposed to be called from foreground thread
+    // and only call if SchedulingAfterWorkCallbackFailed() returns false.
+    CHIP_ERROR RunAfterWorkCallback()
+    {
+        // Ensure that this function is being called from main Matter thread
+        assertChipStackLockedByCurrentThread();
```

Comment:
Actually regarding `IsCancelled` there's two kinds of cancel: 1) the kind called from the main thread, which tries to avoid scheduling part C and aborting, and 2) the kind from the work function itself, which also tries to avoid scheduling part C. In the former case, presumably the main thread has already handled what it needs to, and in the latter case, it's already not scheduling part C, so failure to do so isn't relevant. So I don't think we need to do anything for cancel, we just need to ensure the main thread is cleaning up and doing any responses (if it desires) in all cases.